File: utils/validation.py
"""Input validation utilities for web endpoints and CLI."""

import re
from typing import Any, Dict, List, Optional, Union
from pathlib import Path
import json


class ValidationError(ValueError):
    """Custom exception for validation errors."""
    pass


def validate_positive_int(value: Any, field_name: str = "value", max_value: Optional[int] = None) -> int:
    """Validate that a value is a positive integer."""
    try:
        int_val = int(value)
    except (ValueError, TypeError):
        raise ValidationError(f"{field_name} must be a valid integer")
    
    if int_val <= 0:
        raise ValidationError(f"{field_name} must be positive")
    
    if max_value is not None and int_val > max_value:
        raise ValidationError(f"{field_name} must not exceed {max_value}")
    
    return int_val


def validate_non_negative_float(value: Any, field_name: str = "value") -> float:
    """Validate that a value is a non-negative float."""
    try:
        float_val = float(value)
    except (ValueError, TypeError):
        raise ValidationError(f"{field_name} must be a valid number")
    
    if float_val < 0:
        raise ValidationError(f"{field_name} must be non-negative")
    
    return float_val


def validate_string(value: Any, field_name: str = "value", max_length: int = 1000, allow_empty: bool = False) -> str:
    """Validate that a value is a safe string."""
    if not isinstance(value, str):
        raise ValidationError(f"{field_name} must be a string")
    
    if not allow_empty and not value.strip():
        raise ValidationError(f"{field_name} cannot be empty")
    
    if len(value) > max_length:
        raise ValidationError(f"{field_name} must not exceed {max_length} characters")
    
    # Basic XSS prevention - reject strings with script tags
    if re.search(r'<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>', value, re.IGNORECASE):
        raise ValidationError(f"{field_name} contains unsafe content")
    
    return value.strip()


def validate_file_path(value: Any, field_name: str = "path", must_exist: bool = False) -> Path:
    """Validate that a value is a safe file path."""
    if not isinstance(value, (str, Path)):
        raise ValidationError(f"{field_name} must be a valid path")
    
    try:
        path = Path(value).resolve()
    except (ValueError, OSError):
        raise ValidationError(f"{field_name} is not a valid path")
    
    # Prevent directory traversal attacks
    if '..' in str(path) or str(path).startswith('/'):
        raise ValidationError(f"{field_name} contains unsafe path components")
    
    if must_exist and not path.exists():
        raise ValidationError(f"{field_name} does not exist")
    
    return path


def validate_json_dict(value: Any, field_name: str = "data", required_keys: Optional[List[str]] = None) -> Dict[str, Any]:
    """Validate that a value is a valid JSON dictionary."""
    if not isinstance(value, dict):
        raise ValidationError(f"{field_name} must be a dictionary")
    
    if required_keys:
        missing_keys = [key for key in required_keys if key not in value]
        if missing_keys:
            raise ValidationError(f"{field_name} missing required keys: {missing_keys}")
    
    return value


def validate_choice(value: Any, choices: List[str], field_name: str = "value") -> str:
    """Validate that a value is one of the allowed choices."""
    if value not in choices:
        raise ValidationError(f"{field_name} must be one of: {choices}")
    
    return value


def sanitize_agent_id(agent_id: Any) -> int:
    """Sanitize and validate agent ID."""
    return validate_positive_int(agent_id, "agent_id", max_value=10000)


def sanitize_proposal_id(proposal_id: Any) -> int:
    """Sanitize and validate proposal ID."""
    return validate_positive_int(proposal_id, "proposal_id", max_value=100000)


def sanitize_simulation_steps(steps: Any) -> int:
    """Sanitize and validate simulation steps."""
    return validate_positive_int(steps, "steps", max_value=10000)


def sanitize_token_amount(amount: Any) -> float:
    """Sanitize and validate token amounts."""
    validated = validate_non_negative_float(amount, "token_amount")
    if validated > 1_000_000_000:  # Reasonable upper bound
        raise ValidationError("token_amount exceeds maximum allowed value")
    return validated

File: lib/utils/validation.ts
// Input validation utilities
// Port from utils/validation.py

/**
 * Custom validation error
 */
export class ValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ValidationError';
  }
}

/**
 * Validate that a value is a positive integer
 */
export function validatePositiveInt(
  value: any,
  fieldName: string = 'value',
  maxValue?: number
): number {
  const intVal = parseInt(value, 10);

  if (isNaN(intVal)) {
    throw new ValidationError(`${fieldName} must be a valid integer`);
  }

  if (intVal <= 0) {
    throw new ValidationError(`${fieldName} must be positive`);
  }

  if (maxValue !== undefined && intVal > maxValue) {
    throw new ValidationError(`${fieldName} must not exceed ${maxValue}`);
  }

  return intVal;
}

/**
 * Validate that a value is a non-negative number
 */
export function validateNonNegativeFloat(
  value: any,
  fieldName: string = 'value'
): number {
  const floatVal = parseFloat(value);

  if (isNaN(floatVal)) {
    throw new ValidationError(`${fieldName} must be a valid number`);
  }

  if (floatVal < 0) {
    throw new ValidationError(`${fieldName} must be non-negative`);
  }

  return floatVal;
}

/**
 * Validate that a value is a safe string
 */
export function validateString(
  value: any,
  fieldName: string = 'value',
  maxLength: number = 1000,
  allowEmpty: boolean = false
): string {
  if (typeof value !== 'string') {
    throw new ValidationError(`${fieldName} must be a string`);
  }

  if (!allowEmpty && !value.trim()) {
    throw new ValidationError(`${fieldName} cannot be empty`);
  }

  if (value.length > maxLength) {
    throw new ValidationError(
      `${fieldName} must not exceed ${maxLength} characters`
    );
  }

  // Basic XSS prevention - reject strings with script tags
  const scriptRegex = /<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/i;
  if (scriptRegex.test(value)) {
    throw new ValidationError(`${fieldName} contains unsafe content`);
  }

  return value.trim();
}

/**
 * Validate that a value is a valid JSON object
 */
export function validateJsonDict(
  value: any,
  fieldName: string = 'data',
  requiredKeys?: string[]
): Record<string, any> {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new ValidationError(`${fieldName} must be an object`);
  }

  if (requiredKeys) {
    const missingKeys = requiredKeys.filter((key) => !(key in value));
    if (missingKeys.length > 0) {
      throw new ValidationError(
        `${fieldName} missing required keys: ${missingKeys.join(', ')}`
      );
    }
  }

  return value;
}

/**
 * Validate that a value is one of the allowed choices
 */
export function validateChoice<T extends string>(
  value: any,
  choices: readonly T[],
  fieldName: string = 'value'
): T {
  if (!choices.includes(value)) {
    throw new ValidationError(
      `${fieldName} must be one of: ${choices.join(', ')}`
    );
  }

  return value as T;
}

/**
 * Sanitize and validate agent ID
 */
export function sanitizeAgentId(agentId: any): number {
  return validatePositiveInt(agentId, 'agent_id', 10000);
}

/**
 * Sanitize and validate proposal ID
 */
export function sanitizeProposalId(proposalId: any): number {
  return validatePositiveInt(proposalId, 'proposal_id', 100000);
}

/**
 * Sanitize and validate simulation steps
 */
export function sanitizeSimulationSteps(steps: any): number {
  return validatePositiveInt(steps, 'steps', 10000);
}

/**
 * Sanitize and validate token amounts
 */
export function sanitizeTokenAmount(amount: any): number {
  const validated = validateNonNegativeFloat(amount, 'token_amount');
  if (validated > 1_000_000_000) {
    throw new ValidationError('token_amount exceeds maximum allowed value');
  }
  return validated;
}

/**
 * Validate percentage value (0-1 or 0-100)
 */
export function validatePercentage(
  value: any,
  fieldName: string = 'percentage',
  asDecimal: boolean = true
): number {
  const floatVal = validateNonNegativeFloat(value, fieldName);
  const maxValue = asDecimal ? 1 : 100;

  if (floatVal > maxValue) {
    throw new ValidationError(
      `${fieldName} must be between 0 and ${maxValue}`
    );
  }

  return floatVal;
}

/**
 * Validate range (min, max)
 */
export function validateRange(
  min: any,
  max: any,
  fieldName: string = 'range'
): [number, number] {
  const minVal = validateNonNegativeFloat(min, `${fieldName}_min`);
  const maxVal = validateNonNegativeFloat(max, `${fieldName}_max`);

  if (minVal > maxVal) {
    throw new ValidationError(`${fieldName} min must be less than max`);
  }

  return [minVal, maxVal];
}

/**
 * Validate array of specific type
 */
export function validateArray<T>(
  value: any,
  fieldName: string = 'array',
  validator?: (item: any, index: number) => T,
  minLength?: number,
  maxLength?: number
): T[] {
  if (!Array.isArray(value)) {
    throw new ValidationError(`${fieldName} must be an array`);
  }

  if (minLength !== undefined && value.length < minLength) {
    throw new ValidationError(
      `${fieldName} must have at least ${minLength} items`
    );
  }

  if (maxLength !== undefined && value.length > maxLength) {
    throw new ValidationError(
      `${fieldName} must have at most ${maxLength} items`
    );
  }

  if (validator) {
    return value.map((item, index) => validator(item, index));
  }

  return value;
}

/**
 * Validate email format (basic)
 */
export function validateEmail(value: any, fieldName: string = 'email'): string {
  const str = validateString(value, fieldName, 255);
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;

  if (!emailRegex.test(str)) {
    throw new ValidationError(`${fieldName} must be a valid email address`);
  }

  return str;
}

/**
 * Validate URL format
 */
export function validateUrl(value: any, fieldName: string = 'url'): string {
  const str = validateString(value, fieldName, 2000);

  try {
    new URL(str);
  } catch {
    throw new ValidationError(`${fieldName} must be a valid URL`);
  }

  return str;
}

/**
 * Sanitize HTML to prevent XSS
 */
export function sanitizeHtml(value: string): string {
  return value
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#x27;')
    .replace(/\//g, '&#x2F;');
}

/**
 * Validate boolean value
 */
export function validateBoolean(
  value: any,
  fieldName: string = 'value'
): boolean {
  if (typeof value === 'boolean') {
    return value;
  }

  if (typeof value === 'string') {
    const lower = value.toLowerCase();
    if (lower === 'true' || lower === '1') return true;
    if (lower === 'false' || lower === '0') return false;
  }

  if (typeof value === 'number') {
    if (value === 1) return true;
    if (value === 0) return false;
  }

  throw new ValidationError(`${fieldName} must be a boolean value`);
}
